### PR TITLE
Devx 903 Authenticating with BI

### PR DESCRIPTION
**Pull request by zachdain-bi**

### CODE REVIEW

File: docs/using-bi-for-auth.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+title: Using Beyond Identity for Authentication
+sidebar_position: 2
+---
+In this guide we will show you how to create a tenant, an application, configure the application, and how to integrate with the Beyond Identity Embedded SDK
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:5]: "In this guide, we will show you how to create a tenant and and application, as well as how to configure that application in order to authorize and authenticate against Beyond Identity"

File: docs/using-bi-for-auth.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+Create a tenant via the self signup form:
+
+https://console-us.beyondidentity.com/signup
+
+After creating a tenant, you will be redirected to an Admin Console for the newly created tenant.
+
+### Create a new Realm
+
+Click the drop down on the top left corner and add a new realm. [[Screenshot](./screenshots/NewRealm.jpg)]
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:17]: I would add this above for context (from the other doc):

"Realms are unique administrative domains within a tenant. All new tenants have a default realm called Beyond Identity Admin which should not be used to configure for delegate IDP purposes."

File: docs/using-bi-for-auth.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+### Create a new Realm
+
+Click the drop down on the top left corner and add a new realm. [[Screenshot](./screenshots/NewRealm.jpg)] 
+
+### Create an Application
+
+- Click 'Applications' in the left menu
+- Click 'Create app'
+- Ensure to fill out values for display name, client type, and redirect uris
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:23]: I would expand this a bit more:
- Fill out the display name with whatever you want to name this app
- Select the client type:
  - (describe the client types down here)
- Set the redirect_uri to be a URL where you want to receive the authorization code and state. This can be either:
  - An app scheme or Universal URL / App Link if you're implementing this in a native application
  - A URL to any page in your web application

File: docs/using-bi-for-auth.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+[[Created App Screenshot](./screenshots/AppCreated.jpg)]
+
+### Create a test identity
+
+Before users can start authenticating with Beyond Identity, they must be provisioned in the Beyond Identity Directory. As Auth0 does not support SCIM, users must be manually provisioned using the Beyond Identity Admin Portal or using the [Create User](ref:createuser) API. See the Admin Portal video tutorial that shows how to navigate to the directory area of the admin portal. 
+
+https://www.beyondidentity.com/resources/beyond-identity-admin-console-overview
+
+- In the 'Admin Portal' under the 'Directory' tab, select 'Add User'.
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:39]: "Admin Portal" -> "Beyond Identity Console"

Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:39]: "Directory" -> "Identities"

Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:39]: "Add User" -> "Add Identity"

File: docs/using-bi-for-auth.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+- Enter the following values:
+
+   - External ID: user_id in Auth0, for example: oidc | Beyond-Identity | <email_address>
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:43]: Remove this line

File: docs/using-bi-for-auth.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+   - Email: '<email_address>
+
+   - Username: '<email_address>
+
+   - Display Name: '<Full_Name>
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:49]: "Display Name" -> "Name"

File: docs/using-bi-for-auth.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+### Craft your authorize URL
+
+A full authorization request url has additional parameters that we need to account for
+
+```
+https://authorization-server.com/auth?
+response_type=code
+&client_id=29352735982374239857
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:58]: client_id=<client_id_from_application>

File: docs/using-bi-for-auth.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+&redirect_uri=https://example-app.com/callback
+&scope=create+delete
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:60]: scope=openid

File: docs/using-bi-for-auth.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+   - Username: '<email_address>
+
+   - Display Name: '<Full_Name>
+
+### Craft your authorize URL
+
+A full authorization request url has additional parameters that we need to account for
+
+```
+https://authorization-server.com/auth?
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:56]: Use this URL: 

https://auth-us.beyondidentity.com/v1/tenants/<tenant_id>/realms/<realm_id>/applications/<application_id>/authorize

I would also take a screenshot of exactly where they can get that URL from under application: 
image

File: docs/using-bi-for-auth.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+&client_id=29352735982374239857
+&redirect_uri=https://example-app.com/callback
+&scope=create+delete
+&state=xcoivjuywkdkhvusuye3kch
+```
+
+An application that is configured as public will need additional query parameters:
+
+```
+code_verifier=[random string]
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:67]: code_verifier is not part of the /authorize request. The rest are fine.

**Review by ismailmustafa — CHANGES_REQUESTED** [review]

File: docs/using-bi-for-auth.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  - Confidential clients are applications that are able to securely authenticate with the authorization server, for example being able to keep their registered client secret safe.
+
+  - Public clients are unable to use registered client secrets, such as applications running in a browser or on a mobile device.
+
+- Set the redirect_uri to be a URL where you want to receive the authorization code and state. This can be either:
+  - An app scheme or Universal URL / App Link if you're implementing this in a native application
+  - A URL to any page in your web application
+
+- After the protocol configuration is finished, modify your authenticator config
+  - [Authenticator Config Reference](https://developer-docs-git-v1-beyondidentity.vercel.app/docs/v1/platform-overview/authenticator-config#embedded)
```
Comment (Will5) [inline docs/using-bi-for-auth.md:36]: I don't think this is the final url. As per Ismail, we should use relative paths here, just so it works in any deployed environment.

File: docs/using-bi-for-auth.md [hunk 12]
```diff
@@ -49,9 +49,9 @@ https://www.beyondidentity.com/resources/beyond-identity-admin-console-overview
 
    - Email: <email_address>
 
-   - Username: <email_address>
+   - Username: <user_name>
 
-   - Name: <name>
+   - Name: <full_name>
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:54]: Let's make sure these match against what the API expects: https://developer-docs-git-v1-beyondidentity.vercel.app/api/v1#tag/Identities/operation/CreateIdentity

File: docs/getting-started.md [hunk 13]
```diff
@@ -9,17 +9,6 @@ In this guide we will show you how to create a tenant and connect the Beyond Ide
 ### Create a Tenant
 Navigate here to [Create a Tenant](https://www.beyondidentity.com/developers/signup).
 
-Once your tenant has been created with Beyond Identity, you can continue to create an API token.
```
Comment (ismailmustafa) [inline docs/getting-started.md:12]: Why is this deleted?

File: docs/kotlin-sdk/kotlin-sdk.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
+---
+title: Overview
+sidebar_position: 1
+---
+
+The Embedded SDK is a holistic SDK solution offering the entire experience embedded in your product. Users will not need to download the Beyond Identity Authenticator. A set of functions are provided to you through the `Embedded` namespace. This SDK supports OIDC and OAuth2.
```
Comment (ismailmustafa) [inline docs/kotlin-sdk/kotlin-sdk.md:6]: Let's rebase so this isn't part of the diff

File: docs/using-bi-for-auth.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+https://console-us.beyondidentity.com/signup
+
+After creating a tenant, you will be redirected to an Admin Console for the newly created tenant.
+
+### Create a new Realm
+
+Realms are unique administrative domains within a tenant. All new tenants have a default realm called Beyond Identity Admin which should not be used to configure for delegate IDP purposes.
+
+Click the drop down on the top left corner and add a new realm. [[Screenshot](./screenshots/NewRealm.jpg)]
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:19]: Can we inline pictures whenever possible?

File: docs/using-bi-for-auth.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+  - A URL to any page in your web application
+
+- After the protocol configuration is finished, modify your authenticator config
+  - [Authenticator Config Reference](https://developer-docs-git-v1-beyondidentity.vercel.app/docs/v1/platform-overview/authenticator-config#embedded)
+
+[[Created App Screenshot](./screenshots/AppCreated.jpg)]
+
+### Create a test identity
+
+Before users can start authenticating with Beyond Identity, they must be provisioned in the Beyond Identity Directory. As Auth0 does not support SCIM, users must be manually provisioned using the Beyond Identity Admin Portal or using the [Create User](https://developer.beyondidentity.com/api/create-user) API. See the Admin Portal video tutorial that shows how to navigate to the directory area of the admin portal.
```
Comment (ismailmustafa) [inline docs/using-bi-for-auth.md:42]: I know I asked you to copy/paste this, but let's modify it to suite the context.

Before users can start authenticating with Beyond Identity, they must be provisioned in Beyond Identity.

The rest is very specific to an integration with Auth0. Let's also remove the link to the video overview of the admin console as that highlights the old admin portal. The new console is going to be called console, or "Beyond Identity Console".

**Review by ismailmustafa — CHANGES_REQUESTED** [review]

File: docs/using-bi-for-auth.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+&redirect_uri=<redirect_uri from application>
+&scope=openid
+&state=<state>
+```
+
+An application that is configured as public will need additional query parameters:
+
+```
+&code_challenge=[SHA256 hash of code_verifier]
+&code_challenge_method=[S256|plain]
```
[the remark this answers is not in the document]
Comment (ismailmustafa) [reply docs/using-bi-for-auth.md:73]: I don't think we do. We should hardcode this to "S256"

File: docs/using-bi-for-auth.md (hunk 12 above)
Comment (zachdain-bi) [reply docs/using-bi-for-auth.md:54]: I was referring to the GUI at that point. I just linked it as an option. Could word it better

File: docs/getting-started.md (hunk 13 above)
Comment (zachdain-bi) [reply docs/getting-started.md:12]: I thought we didn't need this anymore? I can put it back

**Review by ismailmustafa — CHANGES_REQUESTED** [review]
Looks good. A couple of things.

- Can you merged this into v1 instead of the android branch since this is useful for swift as well.
- Can you try inlining the images so that users don't have to click on it?

**Comment by zachdain-bi** [comment]
Alrighty, I inlined the images, rebased, and cleaned up the files that weren't mine

**Review by ismailmustafa — APPROVED** [review]
LGTM. We can get this in and update/expose the integration doc in another PR.
